features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5XPalGattClient.cpp

```diff
@@ -0,0 +1,1616 @@
+                        sizeof(uint16_t)
+                    )
+                };
+                return result;
+            }
+
+            const services_array_t &services;
+        };
+
+        terminate(CustomAttReadByGroupTypeResponse(services));
+    }
+
+    /**
+     * Handle discovery of services with a 128 UUID.
+     *
+     * Handle of the services discovered will be stored locally then the
+     * procedure sequentially initiate a read request of each of these attribute
+     * handle to acquire the value of the UUID of the service.
+     *
+     * The procedure ends once all the informations initially sent by the peer
+     * has been reconstructed and forwarded to the registered client handler.
+     */
+    void handle_128bit_services_discovered(const services_array_t &services)
+    {
+        response = new(std::nothrow) packed_discovery_response_t[services.size()];
+        if (!response) {
+            abort();
+            return;
+        }
+
+        count = services.size();
+        idx = 0;
+        for (size_t i = 0; i < count; ++i) {
+            uint8_t *it = &response[i][0];
+            u16_to_stream(it, services[i].handle_range.start_handle);
+            u16_to_stream(it, services[i].handle_range.end_handle);
+        }
+
+        read_service_uuid();
+    }
+
+    /**
+     * Initiate the read request of the next service attribute handle.
+     */
+    void read_service_uuid(void)
+    {
+        // note: use read multiple once different mtu size are supported ?
+        uint16_t attribute_handle = bytes_to_u16(&response[idx][0]);
+
+        uint32_t err = sd_ble_gattc_read(connection_handle, attribute_handle, 0);
+        if (err) {
+            abort();
+        }
+    }
+
+    /**
+     * Handle reception of a service (long) UUID.
+     */
+    void handle_uuid_read(const ble_gattc_evt_t &evt)
+    {
+        if (evt.gatt_status != BLE_GATT_STATUS_SUCCESS) {
+            terminate(AttErrorResponse(
+                AttributeOpcode::READ_REQUEST,
+                convert_sd_att_error_code(evt.gatt_status)
+            ));
+            return;
+        }
+
+        const ble_gattc_evt_read_rsp_t &rsp = evt.params.read_rsp;
+
+        uint16_t expected_handle = bytes_to_u16(&response[idx][0]);
+
+        if (rsp.handle != expected_handle || rsp.offset != 0 || rsp.len != 16) {
+            abort();
+        }
+
+        memcpy(&response[idx][4], rsp.data, rsp.len);
+
+        ++idx;
+
+        if (idx == count) {
+            terminate(SimpleAttReadByGroupTypeResponse(
+                sizeof(packed_discovery_response_t),
+                make_const_ArrayView(
+                    reinterpret_cast<const uint8_t*>(response),
+                    count * sizeof(packed_discovery_response_t)
+                ))
+            );
+        } else {
+            read_service_uuid();
+        }
+    }
+
+    typedef uint8_t packed_discovery_response_t[20];
```

Comment:
What is the cause for have 20 here?

Reply:
Attribute data present in the read by group type response will be 20 bytes long: 
- Attribute handle: 2 bytes 
- End group handle: 2 bytes 
- Attribute value (128bits UUID): 16 bytes 

I will clarify the code.